### DISCUSSION TITLE
Datetime picker with the current date pre-selected (as default)

### DISCUSSION THREAD

**Discussion by s-wel** [Avaiga/taipy#2361] [Q&A] [answered]
Hi, I have a datetime picker that should show the current datetime when a user visits my page. However, it seems the selected default datetime is always the one when the server was started. What is the correct way of calculating the current datetime (timestamp) for the default value?

My _page.py_ file has this code atm:
```
start_time = datetime.now() - timedelta(days=30)
end_time = datetime.now()
```

And my page.md file has this tag:
`<|{start_time}|date|with_time|format=dd/MM/yyyy HH:mm:ss|>`

**Comment by jrobinAV** [comment, marked as answer]
Hi @s-wel,

Your date visual element is bound to the `state.start_time` variable, initialized at the application start. Then, your variable is never changed, so its value is always the same.

I am not sure I understand your use case.

## 1. You want to initialize your variable at the user session creation

You can use the [`on_init` callback](https://docs.taipy.io/en/latest/refmans/reference/pkg_taipy/pkg_gui/Gui/#taipy.gui.Gui.on_init) to initialize state variables when a new user connects, such as:
```python
from datetime import datetime, timedelta

from taipy.gui import Gui, Markdown

def on_init(state):
    state.current_time = datetime.now()

if __name__ == "__main__":
    current_time = datetime.now()
    gui = Gui(Markdown("<|{current_time}|date|with_time|format=dd/MM/yyyy HH:mm:ss|>"))
    gui.run()
```
With this piece of code, the `state.start_time` variable is set with the current date and time when the user connects (at the session creation). So, two users (or two different browsers) won't have the same value in its state. However, the variable won't be updated once the session is created. 

## 2. You want to display the current time in your application
In this case, you need to trigger the variable change when you want it to be updated (for example, every second).
You can use a [taipy scheduler](https://docs.taipy.io/en/latest/userman/advanced_features/scheduling/), and the [broadcast_change](https://docs.taipy.io/en/latest/refmans/reference/pkg_taipy/pkg_gui/Gui/#taipy.gui.Gui.broadcast_change) function, such as:
```python
from datetime import datetime
import taipy as tp
from taipy.gui import Gui, Markdown

def refresh(gui):
    gui.broadcast_change("current_time", datetime.now())

if __name__ == "__main__":
    current_time = datetime.now()
    gui = Gui(Markdown("<|{current_time}|date|with_time|format=dd/MM/yyyy HH:mm:ss|>"))
    tp.Scheduler.every(1).seconds.do(refresh, gui)
    tp.Scheduler.start(1)
    gui.run()

```
Note that the scheduler functionality is part of Taipy Enterprise Edition.
Here is the result:

https://github.com/user-attachments/assets/99480eeb-0aac-45a4-825c-d320a50fb460

**Comment by s-wel** [comment]
The first case worked for me. Thanks.
